### DISCUSSION TITLE
Content Warning Module

### DISCUSSION THREAD

**Discussion by DrewBrasher** [OrchardCMS/OrchardCore#13643] [Ideas]
I would like to create a module that would add a content warning (CW) feature. People use this type of feature a lot on platforms like Discord and Mastodon to hide content that some people might not want to see until they select that they want to see it. I think a feature like this would be especially useful in a blog site.

I'm trying to think about the best way to implement something like this so if anyone has any ideas let me know.

Some things I'm considering are:

- When you enable the feature, should that add the option to put a CW on every content type or should there be a settings page where you specify which content types have the option for a CW?
- I think it would be good if entire content items could have a CW but also a way to just hide sections of text. Discord does this when you enclose the text in double pipes (||). Maybe this could be a shortcode in the OC module.
- Maybe there should be a way for users to turn off CW for themselves if they don't care about that and don't want to have to click to see content.

**Update on 2024-05-31:** I've started building this. The repository is <https://github.com/DrewBrasher/OrchardCoreModules>

**Comment by Piedone** [comment]
Late to the party but I just noticed this topic.

Are we talking about something like an NSFW warning?

I'd handle this on the client side: e.g. add a blur overlay onto the content item's `div` with a button to show the content (and that'd also save it to local storage that you already set that particular item to show).

Otherwise, I'd do this as a content part with an editor checkbox to add a CW to the whole item. Otherwise, perhaps a shortcode to use it for just selected pieces of text/body would work?

**Reply by DrewBrasher** [reply]
Yes, I'm talking about something that could be used for NSFW warning as well as other things like movie spoilers and trigger warnings.

I think doing both the content part and the short code would be good because sometimes just a small piece for text needs a CW and sometimes the whole content might need it.

One thing I need to look into more is how to handle when people are using accessibility features like a screen reader.  In that case I think it should read the warning and wait to for some kind of confirmation before reading the content. I'm not sure how this is handled on other platforms that use content warnings.

**Comment by ns8482e** [comment]
This can be achieved by creating a Part with ContentPicker field. Logic could be similar to FormPart where when a Part is attached, a ContentDriver will add a wrapper shape around the content. In wrapper you could render selected content item from content picker and show/hide/blur the original content using css/js

**Reply by DrewBrasher** [reply]
Thanks! Adding a wrapper shape sounds like a good way to do it. I will take a look at the FormPart.

**Reply by DrewBrasher** [reply]
I tried doing it as a wrapper like the FormPart but then I realized that I don't actually want to wrap the whole content item. I think the TitlePart should not be hidden. I'm trying to figure out how I could wrap everything except the title part but I'm not sure that's possible.

**Reply by Piedone** [reply]
I can't check now, but I think it might be possible to wrap content zones (it certainly is possible to wrap global - layout - zones), like `Content` (where the body of a content item is displayed). Check out with [Shape Tracing](https://github.com/Lombiq/Helpful-Extensions?tab=readme-ov-file#shape-tracing-helpful-extensions) if there's an existing wrapper.

If not, then you can run the debugger with an `IShapeDisplayEvents` implementation to see if `DisplayingAsync` is called for a content zone. Even if not, it'll be called for the `Content` shape, at which point you'll be able to access the shape to be displayed in its `Content` zone in the `Content` property (yes, both are "Content") and add a wrapper on it directly.

**Reply by DrewBrasher** [reply]
@Piedone Thank you, I'll try that.

**Reply by DrewBrasher** [reply]
After enabling shape tracing and looking at the rendered html closer I think the wrapper is fine the way it is. It is already not wrapping the title in the detail view. I think I just got confused because it does wrap the whole thing including the title in the summary view but now I'm thinking that the summary view should not even be wrapped in a warning because only the title and subtitle are displayed anyway.

The issue in the detail view is that the script and stylesheet elements don't get rendered properly for some reason. they are rendered like this:
```
<style asp-src="~/DrewBrasher.OrchardCore.ContentWarning/ContentWarning.css"></style>
<script asp-src="~/DrewBrasher.OrchardCore.ContentWarning/ContentWarning.js" at="Foot"></script>
```
The `asp-src=` is not getting processed in the detail view but it is in the summary view.

**Reply by DrewBrasher** [reply]
I think for the summary view, I'll just append the warning message to the end of the summary.

**Reply by DrewBrasher** [reply]
I fixed the issue with the ContentWarning js and css files not getting loaded in the details view by registering them with the resource manager in the content driver instead of referencing them in the wrapper view.

**Comment by DrewBrasher** [comment]
Here is an article that describes the kind of behavior I would like to implement including accessibility consideration: 
https://www.aaron-gustafson.com/notebook/considering-content-warnings-in-html/

and here is the codepen from that article: https://codepen.io/aarongustafson/pen/QWZpqPe

**Reply by DrewBrasher** [reply]
Are only certain attributes allowed to be rendered in a shortcode? I'm trying to add a data- attribute to the output and it just gets stripped out. Here is how I'm trying to add it:
```
    public ValueTask<string> EvaluateAsync(string identifier, Arguments arguments, string content, Context context)
    {
        var warning = arguments.NamedOrDefault("warning");
        content = $"<span data-content-warning=\"{warning}\">{content}</span>";
        return new ValueTask<string>(content);
    }
```

But if I use the shortcode like this:
```
I loved when [content_warning 'Star Trek Discovery spoiler'] this one thing happened [/content_warning].
```

it outputs this html:
```
I loved when <span> this one thing happened </span>.
```

**Reply by Piedone** [reply]
While I don't know the solution from the top of my head, I'd suggest you debug and play with a shortcode provider built into OC, with the OC source, and see what you can find out.

**Reply by DrewBrasher** [reply]
I stepped through the Orchard Source code using the asset_url shortcode and found that the reason it was not getting rendered was because the markdown part of the blog post was set to sanitize html. I unchecked that and it works.

**Reply by hishamco** [reply]
As expected

**Reply by Piedone** [reply]
Sanitization should probably happen after shortcodes are rendered though, since the point of shortcodes is to provide alternative dynamic behavior that you otherwise can't achieve with a restricted editor.

**Reply by DrewBrasher** [reply]
> Sanitization should probably happen after shortcodes are rendered though, since the point of shortcodes is to provide alternative dynamic behavior that you otherwise can't achieve with a restricted editor.

@Piedone Is there already an open issue for this or should one be created?

**Reply by Piedone** [reply]
I don't think there is one, please open one.

**Reply by DrewBrasher** [reply]
Opened issue https://github.com/OrchardCMS/OrchardCore/issues/16226
